ntfs_forensics/src/ntfsindxtable.cpp

```diff
@@ -98,59 +53,199 @@ void populateIndexRow(osquery::Row& r,
-    trailofbits::Device* d = NULL;
-    trailofbits::Partition* p = NULL;
-    try {
-      d = new trailofbits::Device(dev);
-      p = new trailofbits::Partition(*d, partition);
-    } catch (std::runtime_error&) {
-      delete p;
-      delete d;
+  std::unordered_set<std::uint32_t> partition_constraints;
+  for (const auto& partition_str :
+       request.constraints["partition"].getAll(osquery::EQUALS)) {
+    char* null_term_ptr = nullptr;
+    auto partition_number =
+        std::strtoul(partition_str.c_str(), &null_term_ptr, 10);
+
+    if (*null_term_ptr != 0) {
+      VLOG(1) << "Invalid partition specified: " << partition_str
+              << ". Skipping...";
+
       continue;
     }
 
-    trailofbits::DirEntryList entries;
-    trailofbits::FileInfo fileInfo;
-    if (paths.size() == 1) {
-      p->collectINDX(std::string(*(paths.begin())), entries);
-      p->getFileInfo(*(paths.begin()), fileInfo);
-    } else if (inodes.size() == 1) {
-      std::stringstream inode_str;
-      uint64_t inode;
-      inode_str << *inodes.begin();
-      inode_str >> inode;
-      p->collectINDX(inode, entries);
-      p->getFileInfo(inode, fileInfo);
+    partition_constraints.insert(static_cast<std::uint32_t>(partition_number));
+  }
+
+  // Iterate through all devices
+  osquery::QueryData results;
+
+  for (const auto& device_name : device_constraints) {
+    // Make sure the specified device exists
+    auto device_map_it = device_map.find(device_name);
+    if (device_map_it == device_map.end()) {
+      VLOG(1) << "Device " << device_name << " was not found. Skipping...";
+      continue;
     }
 
-    for (auto& entry : entries) {
-      osquery::Row r;
-      populateIndexRow(r, entry, dev, partition, fileInfo.path);
-      results.push_back(r);
+    const auto& available_device_partitions = device_map_it->second;
+
+    // Iterate through all partitions
+    std::unordered_set<std::uint32_t> partition_list;
+    if (partition_constraints.empty()) {
+      partition_list = available_device_partitions;
+    }
+
+    for (const auto& partition_number : partition_list) {
+      try {
+        auto disk_device = std::make_shared<DiskDevice>(device_name);
+        auto disk_partition =
+            std::make_shared<DiskPartition>(disk_device, partition_number);
+
+        // Use the constraint the user has selected to emit the rows
+        if (!inode_constraints.empty()) {
+          generateAndAppendRows(results,
+                                inode_constraints,
+                                disk_partition,
+                                device_name,
+                                partition_number);
+        } else {
+          generateAndAppendRows(results,
+                                path_constraints,
+                                disk_partition,
+                                device_name,
+                                partition_number);
+        }
+
+      } catch (const std::exception&) {
```

Comment:
consider catching a specific exception, rather than all of them. The DiskPartition ctor raises only runtime_error on failure, and that could even be made more specific.

Alternatively refactor the DiskPartition class so that is has a `valid` flag or something and check that after construction and avoid exceptions as return values entirely

Reply:
I plan on removing the exceptions I found in the original implementation